File: TumCampusApp/src/de/tum/in/tumcampusapp/models/Event.java
package de.tum.in.tumcampusapp.models;

import java.util.Date;

import de.tum.in.tumcampusapp.auxiliary.Utils;

/**
 * Event object
 */
public class Event {

	/**
	 * Description, multiline
	 */
	public String description;

	/**
	 * Event end DateTime
	 */
	public Date end;

	/**
	 * Event Facebook-ID
	 */
	public String id;

	/**
	 * Local image, e.g. /mnt/sdcard/tumcampus/events/cache/xy.jpg
	 */
	public String image;

	/**
	 * Event link, e.g. http://www.xyz.de
	 */
	public String link;

	/**
	 * Location, e.g. Munich
	 */
	public String location;

	/**
	 * Name, e.g. PartyX
	 */
	public String name;

	/**
	 * Event start DateTime
	 */
	public Date start;

	/**
	 * New Event
	 * 
	 * <pre>
	 * @param id Event Facebook-ID
	 * @param name Name, e.g. PartyX
	 * @param start Event start DateTime
	 * @param end Event end DateTime
	 * @param location Location, e.g. Munich
	 * @param description Description, multiline
	 * @param link Event link, e.g. http://www.xyz.de
	 * @param image Local image, e.g. /mnt/sdcard/tumcampus/events/cache/xy.jpg
	 * </pre>
	 */
	public Event(String id, String name, Date start, Date end, String location, String description, String link, String image) {

		this.id = id;
		this.name = name;
		this.start = start;
		this.end = end;
		this.location = location;
		this.description = description;
		this.link = link;
		this.image = image;
	}

	@Override
	public String toString() {
		return "id=" + id + " name=" + name + " start=" + Utils.getDateString(start) + " end=" + Utils.getDateString(end) + " location=" + location
				+ " description=" + description + " link=" + link + " image=" + image;
	}
}

File: app/src/main/java/de/tum/in/tumcampusapp/component/ui/ticket/model/Event.kt
package de.tum.`in`.tumcampusapp.component.ui.ticket.model

import android.content.Context
import android.os.Parcel
import android.os.Parcelable
import android.text.format.DateFormat
import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.PrimaryKey
import androidx.room.RoomWarnings
import com.google.gson.annotations.SerializedName
import de.tum.`in`.tumcampusapp.utils.readDateTime
import de.tum.`in`.tumcampusapp.utils.writeDateTime
import org.joda.time.DateTime
import org.joda.time.format.DateTimeFormat

/**
 * Event
 *
 * @param id          Event-ID
 * @param imageUrl    Image url e.g. http://www.tu-film.de/img/film/poster/Fack%20ju%20Ghte.jpg
 * @param title       Title
 * @param description Description
 * @param locality    Locality
 * @param startTime   DateTime
 * @param endTime     DateTime
 * @param eventUrl    Url, e.g. http://www.in.tum.de
 */
@Entity(tableName = "events")
@SuppressWarnings(RoomWarnings.DEFAULT_CONSTRUCTOR)
data class Event(
        @PrimaryKey
        @SerializedName("event")
        var id: Int = 0,
        @SerializedName("file")
        @ColumnInfo(name = "image_url")
        var imageUrl: String? = null,
        var title: String = "",
        var description: String = "",
        var locality: String = "",
        @SerializedName("start")
        @ColumnInfo(name = "start_time")
        var startTime: DateTime = DateTime(),
        @SerializedName("end")
        @ColumnInfo(name = "end_time")
        var endTime: DateTime? = null,
        @SerializedName("link")
        @ColumnInfo(name = "event_url")
        var eventUrl: String = "",
        @ColumnInfo(name = "dismissed")
        var dismissed: Int = 0,
        var kino: Int = -1,
        var news: Int = -1
) : Parcelable, Comparable<Event> {

    // Unsafe calls are only ok because we control writeToParcel().
    // Keep in sync with writeToParcel()!
    constructor(parcel: Parcel) : this(
            id = parcel.readInt(),
            imageUrl = parcel.readString(),
            title = parcel.readString()!!,
            description = parcel.readString()!!,
            locality = parcel.readString()!!,
            startTime = DateTime(parcel.readLong()),
            endTime = parcel.readDateTime(),
            eventUrl = parcel.readString()!!,
            dismissed = parcel.readInt(),
            kino = parcel.readInt(),
            news = parcel.readInt())

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeInt(id)
        parcel.writeString(imageUrl)
        parcel.writeString(title)
        parcel.writeString(description)
        parcel.writeString(locality)
        parcel.writeLong(startTime.millis)
        parcel.writeDateTime(endTime)
        parcel.writeString(eventUrl)
        parcel.writeInt(dismissed)
        parcel.writeInt(kino)
        parcel.writeInt(news)
    }

    fun getFormattedStartDateTime(context: Context): String {
        val date = DateTimeFormat.longDate().print(startTime)
        val pattern = if (DateFormat.is24HourFormat(context)) "H:mm" else "h:mm aa"
        val time = DateTimeFormat.forPattern(pattern).print(startTime)
        return "$date, $time"
    }

    override fun compareTo(other: Event): Int {
        return startTime.compareTo(other.startTime)
    }

    override fun describeContents(): Int {
        return 0
    }

    companion object {

        const val defaultDuration = 7200000 // Milliseconds

        @JvmField var CREATOR = object : Parcelable.Creator<Event> {
            override fun createFromParcel(parcel: Parcel) = Event(parcel)

            override fun newArray(size: Int) = arrayOfNulls<Event?>(size)
        }

    }

}